snuba/datasets/discover.py

```diff
@@ -115,10 +106,31 @@ def match_query_to_table(
                     if event_type == "transaction":
                         return EVENTS
 
-    if len(event_types) == 1 and event_types.pop() == "transaction":
+    if len(event_types) == 1 and "transaction" in event_types:
         return TRANSACTIONS
 
-    return EVENTS
+    if len(event_types) > 0 and "transaction" not in event_types:
+        return EVENTS
+
+    # If we cannot clearly pick a table from the top level conditions, then
+    # inspect the columns requested to infer a selection.
+    has_event_columns = False
+    has_transaction_columns = False
+    for col in query.get_all_ast_referenced_columns():
+        if events_only_columns.get(col.column_name):
+            has_event_columns = True
+        elif transactions_only_columns.get(col.column_name):
+            has_transaction_columns = True
+
+    if has_event_columns and has_transaction_columns:
+        # Impossible query, use the merge table
+        return EVENTS_AND_TRANSACTIONS
+    elif has_event_columns:
+        return EVENTS
+    elif has_transaction_columns:
+        return TRANSACTIONS
+    else:
+        return EVENTS_AND_TRANSACTIONS
```

Comment:
This seems should be an error.

Reply:
Sorry, what does that mean?

Reply:
Nevermind, I forgot that has_event_columns and has_transaction_columns means having columns that are exclusively in one of those tables. A query for the merge table should fall here. Ignore my comment.